### ISSUE TITLE
Scroll position snaps to 0 for a frame when navigating away from a ScrollableGrid or ScrollableList

### ISSUE THREAD

**Issue by mpcomplete** [flutter/flutter#4597]
When you tap a recipe card from the main recipe grid, the scroll position briefly jumps back to the top before switching to the recipe page. I think this has something to do with ScrollableGrid. I've tried changing it to a Block instead, and I don't see the same issue.

**Comment by mpcomplete** [comment]
cc @abarth in case this rings a bell

**Comment by abarth** [comment]
Interesting.  Doesn't ring a bell, but I'm happy to take a look if you'd like.

**Comment by mpcomplete** [comment]
Using a ScrollableList in place of ScrollableGrid exhibits the same spazziness, but a ScrollableViewport does not.

**Comment by abarth** [comment]
Yeah, they share a lot of code.  It's probably in the common code between them.

**Comment by mpcomplete** [comment]
Tracked it down to a new ScrollableState being created the instant of the tap. The new ScrollableState creates a new ScrollableBehavior that is unaware of the content and container extents (until handleExtentsChanged is called the next frame), so you see a flash where the scroll offset is clipped to 0 for these extents.

**Comment by mpcomplete** [comment]
Passing a global scrollableKey to the ScrollableGrid fixed it.

I still think there's a bug here. We shouldn't be creating a new ScrollableState.

**Comment by mpcomplete** [comment]
ScrollableViewport seems to be spared from this bug because its renderer (RenderViewport) has a onPaintOffsetUpdateNeeded callback to immediately adjust the paint offset in response to viewport changes. ScrollableGrid's renderer doesn't have this callback, so there's a frame of paint that happens with a paint offset of 0 before the viewport bounds are updated.

**Comment by mpcomplete** [comment]
You can see this issue with the Grid and List demos in the Gallery app. Anytime you navigate away from a scrolled grid or list (with with a nonzero scroll offset), it snaps back to 0 for a frame before navigating. This includes navigating back.

**Comment by mpcomplete** [comment]
Looking into this more, I'm seeing another problem. ScrollableGrid (and ScrollableList) clamps the scrollOffset to the content's size in its build function. But we don't know the content's size until layout. We then have 1 frame of paint before the grid is rebuilt to accommodate the actual content size calculated in the first layout.

This order seems wrong to me. We shouldn't clamp the scroll offset until we know the actual bounds after the first layout.

**Comment by abarth** [comment]
Yes, that seems wrong.  We should probably pass through the unclamped value and clamp it later once we know how big things are.

**Comment by Hixie** [comment]
The rendering side is supposed to do a synchronous callback during layout with the actual numbers to get the scroll offset updated. Search foc onPaintOffsetUpdateNeeded.

**Comment by mpcomplete** [comment]
There is a synchronous callback (though ScrollableGrid and ScrollableList call it onExtentsChanged). The problem is that it updates the extents, but the paint offset is still based on a Widget's scrollOffset - which has been clamped to 0 at build time. We have to wait for a rebuild (triggered by that synchronous callback) before getting the proper scrollOffset.

**Comment by Hixie** [comment]
the callback is supposed to return the paint offset to actually use, IIRC.

**Comment by mpcomplete** [comment]
The ScrollableList/ScrollableGrid versions don't, but even if they did, it would get overwritten [here](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/virtual_viewport.dart#L131).

**Comment by Hixie** [comment]
Doesn't that line mean that we _are_ updating it before paint, rather than wait for the rebuild?

**Comment by mpcomplete** [comment]
Yes, it's updating. But it's updating based on the widget's scrollOffset, which has been clamped to 0 during the first build. We have to wait for the rebuild for the scrollOffset to be clamped again based on the updated content size.

**Comment by mpcomplete** [comment]
Actually, I think I can paper over this by having onExtentsChanged update the paintOffset after that line (_updatePaintOffset). I might do that to fix this bug, though I still think ClampOverscrolls is a problem.

**Comment by Hixie** [comment]
We shouldn't paper over it, we should find the real fix.

I don't understand why we have to wait to rebuild. Why do we need to clamp during build? Why can't we just proactively do whatever it is we are supposed to do to find the clamped value, in response to the callback?

How does this work for the other scrollable widgets? We should just use that same solution here, if we have something that works and it is appropriate here too.

**Comment by mpcomplete** [comment]
AFAICT the other scrollable widgets paper over this problem in the same way (they update paintOffset in response to onPaintOffsetUpdateNeeded). The first build always has an invalid scrollOffset and paintOffset.

> I don't understand why we have to wait to rebuild. Why do we need to clamp during build? 

I think that's the problem. We shouldn't be clamping during build, but that's how ClampOverscrolls currently works.

**Comment by Hixie** [comment]
Let's change that then. :-)

**Comment by mpcomplete** [comment]
Just spoke to @HansMuller . This is a bit more complicated. There's a circular dependency between the scrollOffset and the content size: LazyBlock needs to know the scrollOffset to decide what to draw, which determines the content size. And as I pointed out above, ClampOverscrolls needs the content size to clamp the scrollOffset. A proper fix might need to do layout in multiple passes (layout to find the content size, clamp the scroll offset, layout again at the new scroll offset, paint).

**Comment by Hixie** [comment]
I wonder if https://github.com/flutter/flutter/issues/4430 would provide a solution. (I haven't thought about this specific case in that context yet.)

**Comment by mpcomplete** [comment]
Found the 2 root causes of this bug. Summarizing here for posterity.

1. The ScrollableGrid is wrapped by a OverscrollIndicator (Android-only). When we navigate away, this indicator animates, which changes its build function. It returns a `new CustomPaint(child: child)` rather than the `child` directly. This causes the Scrollable to be rebuilt and its old ScrollableState to be discarded. This combines with `2` to cause the bug.
2. ScrollableGrid's Scrollable is wrapped by a ClampOverscrolls widget. This widget uses the content size to clamp the scrollOffset. Since we discarded the ScrollableState in `1`, the content size is 0, so we clamp the offset to 0.

Possible fixes:

1. Always return have OverscrollIndicator return a `new CustomPaint`, so the hierarchy never changes. This way, we never rebuild the Scrollable.
2. Redesign ClampOverscrolls to layout the content before clamping the scrollOffset, so we know what size to clamp to.
